Sources/NIOOpenSSL/ByteBufferBIO.swift

```diff
@@ -0,0 +1,289 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import CNIOOpenSSL
+
+
+/// The OpenSSL entry point to write to the `ByteBufferBIO`. This thunk unwraps the user data
+/// and then passes the call on to the specific BIO reference.
+internal func openSSLBIOWriteFunc(bio: UnsafeMutablePointer<BIO>?, buf: UnsafePointer<CChar>?, len: CInt) -> CInt {
+    guard let concreteBIO = bio, let concreteBuf = buf else {
```

Comment:
just a question, why not `guard let bio = bio, buf = buf else {` ? Leaves fewer useless variables in scope

Reply:
As discussed in person, we want to use the old values in the body of the `else` block, which requires we bind to new names.